pulsar-broker/src/main/java/org/apache/pulsar/broker/service/plugin/EntryFilterProvider.java

```diff
@@ -177,12 +180,55 @@ private static EntryFilterWithClassLoader load(EntryFilterMetaData metadata,
             }
             EntryFilter pi = (EntryFilter) filter;
             return new EntryFilterWithClassLoader(pi, ncl);
-        } catch (Exception e) {
+        } catch (Throwable e) {
             if (e instanceof IOException) {
                 throw (IOException) e;
             }
-            log.error("Failed to load class {}", def.getEntryFilterClass(), e);
+            log.error("Failed to load class {}", metadata.getDefinition().getEntryFilterClass(), e);
             throw new IOException(e);
         }
     }
+
+    private NarClassLoader getNarClassLoader(Path archivePath) {
+        return cachedClassLoaders.get(classLoaderKey(archivePath));
+    }
+
+    private NarClassLoader loadNarClassLoader(Path archivePath) {
+        final String absolutePath = classLoaderKey(archivePath);
+        return cachedClassLoaders
+                .computeIfAbsent(absolutePath, narFilePath -> {
+                    try {
+                        final File narFile = archivePath.toAbsolutePath().toFile();
+                        return NarClassLoaderBuilder.builder()
+                                .narFile(narFile)
+                                .parentClassLoader(EntryFilter.class.getClassLoader())
+                                .extractionDirectory(serviceConfiguration.getNarExtractionDirectory())
+                                .build();
+                    } catch (IOException e) {
+                        throw new RuntimeException(e);
+                    }
+                });
+    }
+
+    private static String classLoaderKey(Path archivePath) {
+        return archivePath.toFile().getAbsolutePath();
```

Comment:
is archivePath.toFile().getAbsolutePath() potentially performing some IO operation ? (metadata access?)

probably archivePath.toString() is enough in this context